VSR/Journal: Carefully truncate out-of-view prepares during recovery

sentientwaffle opened the pull request:
Fixes https://github.com/tigerbeetledb/tigerbeetle/issues/783.

### Bug

In the seed reference in the above issue (`14295907088955483715` on `fd94fadf804e6ce612372106093d168fe128e6d7`), the simulator fails in `simulator.zig` on

    assert(replica.status != .recovering_head or fault);

Specifically, replica `1` restarts into `recovering_head` even though storage faults were disabled during its recovery. Slot 21 of the WAL is the problem:

    [warn] (journal): 1: recover_slot: recovered slot=0021 label=@H decision=vsr command=Command.reserved op=21

Case `@H` is:
- the redundant header is a valid prepare, but
- the prepare is reserved.

The prepare is actually present in the log. It is being converted to `reserved` by `recover_slots()`:

    // Discard headers which we are certain do not belong in the current log_view.

### Fix

There are two possible fixes here.
The first is to remove the redundant header at the same time we remove the prepare. The second is to defer the removal, and modify the recovery decision directly.

There was originally a note that this code should precede `prepare_op_max`'s computation. But I think that is actually not right — even if we are pretending those last messages don't exist because they don't belong in our newer view, it doesn't mean that the earlier message should be treated as op-max for the purposes of the recovery decision table.

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

sentientwaffle committed:
VSR/Journal: Carefully truncate out-of-view prepares during recovery

Fixes https://github.com/tigerbeetledb/tigerbeetle/issues/783.

Bug

In the seed reference in the above issue (`14295907088955483715` on `fd94fadf804e6ce612372106093d168fe128e6d7`),
the simulator fails in `simulator.zig` on

    assert(replica.status != .recovering_head or fault);

Specifically, replica `1` restarts into `recovering_head` even though storage faults were disabled during its recovery.
Slot 21 of the WAL is the problem:

    [warn] (journal): 1: recover_slot: recovered slot=0021 label=@H decision=vsr command=Command.reserved op=21

Case `@H` is:
- the redundant header is a valid prepare, but
- the prepare is reserved.

The prepare is actually present in the log. It is being converted to `reserved` by `recover_slots()`:

    // Discard headers which we are certain do not belong in the current log_view.

Fix

There are two possible fixes here.
The first is to remove the redundant header at the same time we remove the prepare.
The second is to defer the removal, and modify the recovery decision directly.

There was originally a note that this code should precede `prepare_op_max`'s computation.
But I think that is actually not right — even if we are pretending those last messages don't exist because they don't belong in our newer view,
it doesn't mean that the earlier message should be treated as op-max for the purposes of the recovery decision table.

matklad commented on src/vsr/journal.zig:
```diff
@@ -1459,9 +1462,8 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
-                    assert(prepare == null);
-                    assert(!journal.prepare_inhabited[slot.index]);
-                    assert(journal.prepare_checksums[slot.index] == 0);
+                    // If `prepare` is non-null, it is being truncated due to the view range.
```
Might split out `.cut_view_range` for tighter asserts?

matklad commented on src/vsr/journal.zig:
```diff
@@ -1303,6 +1279,33 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
                 });
             }
 
+            // Discard headers which we are certain do not belong in the current log_view.
```
Is there some comment somewhere which explains how prepares with a wrong view can arise in the first place?

sentientwaffle replied:
I'll add this explanation:
```
            // (These headers can originate if we join a view, write some prepares from the new
            // view, and then crash before the view_durable_update() finished.)
```

sentientwaffle committed:
VSR/Journal: Explain view-range truncation

sentientwaffle committed:
VSR/Journal: Separate cut cases

This improves logging and allows for stricter assertions.

matklad reviewed (commented):
This looks reasonable to me, but, tbh, I don't yet fully understand this reasoning: 

>But I think that is actually not right — even if we are pretending those last messages don't exist because they don't belong in our newer view, it doesn't mean that the earlier message should be treated as op-max for the purposes of the recovery decision table.

What would be the case where it matters whether we compute `prepare_op_max` before or after cutting out-of-view prepares? 

Is it something like this: 

* `R0` is the primary
* `R0` prepares `op` and crashes
* `R0` restarts, and joints the new view
* `R1`, the new primary sends `op+1` to `R0`
* `R0` writes `op+1`, but not its `log_view`, and crashes again
* `R0` restarts. It wants to truncate `op+1` as out-of-view.
* depending on whether we truncate before or after, `prepare_op_max` might be `op` or `op + 1`
* So, if the header for `op` gets corrupted, we might end up in `@E` or `@F` case (we either trust prepare and replace header locally, or mark slot as faulty)

Does it matter which we choose?

In particular, what if `op+1` is a super trainsient op. THat is, `R1` crashes after sending prepare for `op+1`, so, from the prespective of the cluster, this op just doesn't exist, and `R2`, the next primary, should pick `op` as the head?

sentientwaffle committed:
VSR/Journal: Fix recovery table comment

`a11e314783696651efed8bc7a432c93a84264fad` adjusted this case, but didn't correctly update the comment.
The code table is correct.

sentientwaffle committed:
VSR/Journal: Fix recovery documentation

sentientwaffle committed:
VSR/Journal: Fix recovery test case exhaustive check

The last bit wasn't being checked. The actual table was already exhaustive, so the test still passes.

sentientwaffle commented:
As I was I was writing up an explanation, I finally realized what I was thinking when I wrote this originally:

```
            // It is essential that this is performed before we compute the op_max so that the
            // recovery cases apply correctly.
```

Which means that this code still isn't right. The problem is that `recover_torn_prepare` computes the op-max itself from the headers/prepares. But if this is done _after_ the view-range truncation, then this calculation is incorrect, and might end up truncating a "torn prepare" that wasn't actually torn. Which could then lead to the replica nacking a prepare that it shouldn't nack.

sentientwaffle commented:
Actually nevermind, now I think it was broken before but correct now. :thinking:

sentientwaffle committed:
VSR/Journal: Add recover_torn_prepare explanation to view-range truncation

matklad reviewed (approved):
I think everything makes sense to me know, thanks for extra comments!
